File: frontend/src/services/downloads/download-image.ts
import apiClient from "@/services/api-client";
import { ReturnErrorMessage } from "@/services/api-error-return";
import { updateArtworkVersion } from "@/services/downloads/artwork-version";

import { log } from "@/lib/logger";

import type { APIResponse } from "@/types/api/api-response";
import type { MediaItem } from "@/types/media-and-posters/media-item-and-library";
import type { ImageFile } from "@/types/media-and-posters/sets";

export interface DownloadImageFileForMediaItem_Request {
  media_item: MediaItem;
  image_file: ImageFile;
}

export interface DownloadImageFileForMediaItem_Response {
  result: string;
  updated_at?: number;
}

export const downloadImageFileForMediaItem = async (
  imageFile: ImageFile,
  mediaItem: MediaItem,
  fileName: string
): Promise<APIResponse<DownloadImageFileForMediaItem_Response>> => {
  try {
    const req: DownloadImageFileForMediaItem_Request = {
      image_file: imageFile,
      media_item: mediaItem,
    };
    const response = await apiClient.post<APIResponse<DownloadImageFileForMediaItem_Response>>(
      `/download/image/item`,
      req
    );
    if (response.data.status === "error") {
      throw new Error(
        response.data.error?.message || "Unknown error downloading poster file and updating media server"
      );
    }
    updateArtworkVersion(mediaItem, response.data.data?.updated_at);
    return response.data;
  } catch (error) {
    log(
      "ERROR",
      "API - Media Server",
      "Download and Update",
      `Failed to download file '${fileName}' and update media item '${mediaItem.title}' (TMDB ID: ${mediaItem.tmdb_id}): ${
        mediaItem.rating_key
      }: ${error instanceof Error ? error.message : "Unknown error"}`,
      error
    );
    return ReturnErrorMessage<DownloadImageFileForMediaItem_Response>(error);
  }
};

File: backend/routing/download/download_image.go
package routes_download

import (
	"aura/logging"
	"aura/mediaserver"
	"aura/models"
	"aura/utils"
	"aura/utils/httpx"
	"fmt"
	"net/http"
)

type DownloadImageFileForMediaItem_Request struct {
	ImageFile models.ImageFile `json:"image_file"`
	MediaItem models.MediaItem `json:"media_item"`
}

type DownloadImageFileForMediaItem_Response struct {
	Result    string `json:"result"`
	UpdatedAt int64  `json:"updated_at,omitempty"`
}

// DownloadImageFileForMediaItem godoc
// @Summary      Download Image
// @Description  Download and apply an image for a Media Item in the media server.
// @Tags         Download
// @Accept       json
// @Produce      json
// @Param        req  body      DownloadImageFileForMediaItem_Request  true  "Download Image Request"
// @Security 	 BearerAuth
// @Failure      401  {object}  httpx.UnauthorizedResponse "Unauthorized (only when Auth.Enabled=true)"
// @Success      200           {object}  httpx.JSONResponse{data=DownloadImageFileForMediaItem_Response}
// @Failure      500           {object}  httpx.JSONResponse "Internal Server Error"
// @Router       /api/download/image/item [post]
func DownloadImageFileForMediaItem(w http.ResponseWriter, r *http.Request) {
	ctx, ld := logging.CreateLoggingContext(r.Context(), r.URL.Path)
	logAction := ld.AddAction("Download Image", logging.LevelInfo)
	ctx = logging.WithCurrentAction(ctx, logAction)
	var req DownloadImageFileForMediaItem_Request
	var response DownloadImageFileForMediaItem_Response

	// Parse the request body
	Err := httpx.DecodeRequestBodyToJSON(ctx, r.Body, &req, "Download Image - Decode Request Body")
	if Err.Message != "" {
		httpx.SendResponse(w, ld, response)
		return
	}

	actionValidate := logAction.AddSubAction("Validate Request Body Fields", logging.LevelDebug)
	// Make sure that the mediaItem has the following fields set
	// 1. MediaItem.RatingKey
	// 2. MediaItem.LibraryTitle
	if req.MediaItem.RatingKey == "" || req.MediaItem.LibraryTitle == "" {
		actionValidate.SetError("Invalid Media Item Data", "RatingKey and LibraryTitle are required in Media Item", map[string]any{
			"rating_key":    req.MediaItem.RatingKey,
			"library_title": req.MediaItem.LibraryTitle,
		})
		actionValidate.Complete()
		httpx.SendResponse(w, ld, response)
		return
	}

	// Make sure that the ImageFile has the following fields set
	// 1. ImageFile.ID
	// 2. ImageFile.Type
	if req.ImageFile.ID == "" || req.ImageFile.Type == "" {
		actionValidate.SetError("Invalid Image File Data", "ID and Type are required in Image File", map[string]any{
			"id":   req.ImageFile.ID,
			"type": req.ImageFile.Type,
		})
		actionValidate.Complete()
		httpx.SendResponse(w, ld, response)
		return
	}
	actionValidate.Complete()

	// Make the download and apply the image
	Err = mediaserver.DownloadApplyImageToMediaItem(ctx, &req.MediaItem, req.ImageFile)
	if Err.Message != "" {
		httpx.SendResponse(w, ld, response)
		return
	}

	response.Result = fmt.Sprintf("Successfully downloaded %s", utils.GetFileDownloadName(req.MediaItem.Title, req.ImageFile))
	response.UpdatedAt = req.MediaItem.UpdatedAt
	httpx.SendResponse(w, ld, response)
}
